### PR TITLE
v1.12 backports 2022-10-10

### PR DESCRIPTION
**Pull request by jibi**
* [ ] #21499 -- bpf: always track egress gateway connections (@jibi)
   * all commits manually backported due to some non trivial conflicts

Once this PR is merged, you can update the PR labels via:
```upstream-prs
$ for pr in 21499 ; do contrib/backporting/set-labels.py $pr done 1.12; done
```
or with
```
$ make add-label BRANCH=v1.12 ISSUES=21499
```

### CODE REVIEW

**Comment by jibi** [comment]
/test-backport-1.12

**Comment by pchaigno** [comment]
Any backporting conflicts? If not, why was this carried outside the normal backporting process? We currently have 4 in-flight 1.12 backport PR :scream:

**Comment by jibi** [comment]
> Any backporting conflicts? If not, why was this carried outside the normal backporting process? (We currently have 4 in-flight 1.12 backport PR scream ) If yes, please document slightly_smiling_face

All commits had non trivial conflicts as pointed in the [first backport attempt](https://github.com/cilium/cilium/pull/21610) and I had already [synced offline with the tophat](https://github.com/cilium/cilium/pull/21610#issuecomment-1272278159) about that.

Not sure there's much value in documenting each individual conflict as I literally started from a blank state and manually reapplied the changes

**Comment by pchaigno** [comment]
> Not sure there's much value in documenting each individual conflict as I literally started from a blank state and manually reapplied the changes

Just that would be useful to know when reviewing & when we'll come back to these while trying to understand some buggy behavior.

**Comment by jibi** [comment]
/test-backport-1.12

**Comment by sayboras** [comment]
all 3 jenkins jobs are failing for the same tests, I am not sure if there are some regression here.

Appreciate if someone can chime in.

- K8sDatapathConfig AutoDirectNodeRoutes Check connectivity with automatic direct nodes routes
- K8sDatapathConfig AutoDirectNodeRoutes Check direct connectivity with per endpoint routes
- K8sDatapathConfig Iptables Skip conntrack for pod traffic

https://jenkins.cilium.io/job/Cilium-PR-K8s-1.22-kernel-5.4/780/
https://jenkins.cilium.io/job/Cilium-PR-K8s-1.23-kernel-4.19/583/
https://jenkins.cilium.io/job/Cilium-PR-K8s-1.24-kernel-net-next/626

File: bpf/lib/nat.h [hunk 1]
```diff
@@ -589,6 +589,8 @@ static __always_inline bool snat_v4_needed(struct __ctx_buff *ctx,
                        .saddr = ip4->saddr
                };
 
+	       *from_endpoint = true;
```
Comment (julianwiedmann) [inline bpf/lib/nat.h]: This feels like a risky change. In old code we only set `from_endpoint = true` after checking `!is_reply`, but now it's unconditional.


I believe we should limit this change to the EgressGW path, as follows:

```
target->addr = egress_gw_policy->egress_ip;
target->egress_gateway = true;
if (local_ep)
	*from_endpoint = true;
```

If there's further opportunity for fine-tuning, let's review and push that separately.

Comment (julianwiedmann) [reply bpf/lib/nat.h]: Ah nevermind - just realized that there's no other path in `snat_v4_needed()` that would return true. So even though we set the `from_endpoint` flag a lot earlier, the caller will only look at it under the same conditions as before. Think this is fine then :rocket:.

**Comment by jibi** [comment]
/test-backport-1.12

**Comment by jibi** [comment]
/test-backport-1.12

**Comment by jibi** [comment]
/test-1.23-4.9

**Comment by jibi** [comment]
/test-1.17-4.9

**Comment by jibi** [comment]
/test-1.18-4.9

**Comment by jibi** [comment]
/test-1.19-4.9

**Comment by jibi** [comment]
/test-backport-1.12

**Comment by jibi** [comment]
/test-1.19-4.9

**Review by julianwiedmann — APPROVED** [review]
:crossed_fingers:

**Comment by jibi** [comment]
CI status:

* `ci-aks-1.12` didn't really failed:
```
✅ All 30 tests (224 actions) successful, 0 tests skipped, 1 scenarios skipped.
Error: The operation was canceled.
```

* `test-1.22-4.9` failed because of a known flake (`K8sServicesTest Checks E/W loadbalancing (ClusterIP, NodePort from inside cluster, etc)`)

* `test-gke` and `test-1.23-4.9` are not marked as required

**Pull request merged**